diff --git a/node_modules/node-datachannel/lib/index.cjs b/node_modules/node-datachannel/lib/index.cjs
index 41afd21..d98ab28 100644
--- a/node_modules/node-datachannel/lib/index.cjs
+++ b/node_modules/node-datachannel/lib/index.cjs
@@ -102,7 +102,7 @@ class DataChannelStream extends stream.Duplex {
 }
 
 // createRequire is native in node version >= 12
-const require$1 = module$1.createRequire((typeof document === 'undefined' ? require('u' + 'rl').pathToFileURL(__filename).href : (_documentCurrentScript && _documentCurrentScript.src || new URL('index.cjs', document.baseURI).href)));
+const require$1 = module$1.createRequire((true ? require('u' + 'rl').pathToFileURL(__filename).href : (_documentCurrentScript && _documentCurrentScript.src || new URL('index.cjs', document.baseURI).href)));
 
 const nodeDataChannel = require$1('../build/Release/node_datachannel.node');
 
